File: binary_trees/zig/binarytree.zig
const std = @import("std");

const BinaryTree = struct {
    const Self = @This();

    left: ?*Self = null,
    right: ?*Self = null,

    pub fn init(allocator: std.mem.Allocator) !*Self {
        const bt = try allocator.create(Self);
        bt.* = .{};
        return bt;
    }

    pub fn createBottomUp(depth: u64, allocator: std.mem.Allocator) ?*Self {
        var bt = Self.init(allocator) catch return null;
        if (depth > 0) {
            const d = depth - 1;
            bt.left = Self.createBottomUp(d, allocator);
            bt.right = Self.createBottomUp(d, allocator);
        }
        return bt;
    }

    pub fn countNodes(self: *Self) u64 {
        var sum: u64 = 1;
        if (self.left != null) {
            sum += self.left.?.countNodes();
        }
        if (self.right != null) {
            sum += self.right.?.countNodes();
        }
        return sum;
    }
};

const MIN_DEPTH = 4;

pub fn main(init: std.process.Init) !void {
    const arena = init.arena;
    const allocator = arena.allocator();

    const args = try init.minimal.args.toSlice(allocator);

    if (args.len != 2) {
        std.debug.print("Usage: {s} <depth>", .{args[0]});
        std.process.exit(1);
    }

    const n = try std.fmt.parseInt(u32, args[1], 10);
    const maxDepth = @max(MIN_DEPTH + 2, n);

    var stdout_writer = std.Io.File.stdout().writer(init.io, &.{});
    const output = &stdout_writer.interface;

    {
        const stretchDepth = n + 1;
        const stretchTree = BinaryTree.createBottomUp(stretchDepth, allocator).?;
        const stretchCheck = stretchTree.countNodes();
        try output.print("stretch tree of depth {d}\t check: {d}\n", .{ stretchDepth, stretchCheck });
    }

    var longLivedArea = std.heap.ArenaAllocator.init(init.gpa);
    defer longLivedArea.deinit();
    const longLivedAllocator = longLivedArea.allocator();
    const longLivedTree = BinaryTree.createBottomUp(maxDepth, longLivedAllocator).?;

    var check: u64 = 0;
    var depth: u32 = MIN_DEPTH;
    while (depth <= maxDepth) : (depth += 2) {
        const iterations = @as(usize, 1) << @intCast(maxDepth - depth + MIN_DEPTH);
        check = 0;
        var i: u64 = 1;
        while (i <= iterations) : (i += 1) {
            _ = arena.reset(.retain_capacity);
            const tempTree = BinaryTree.createBottomUp(depth, allocator).?;
            check += tempTree.countNodes();
        }
        try output.print("{d}\t trees of depth {d}\t check: {d}\n", .{ iterations, depth, check });
    }

    try output.print("long lived tree of depth {d}\t check: {d}\n", .{ maxDepth, longLivedTree.countNodes() });
}

File: binary_trees/nim/binarytree.nim
import os, strutils

type
  BinaryTree = ref object
    left, right: BinaryTree

proc newBinaryTree(left, right: BinaryTree): BinaryTree =
  BinaryTree(left: left, right: right)

proc countNodes(bt: BinaryTree): uint =
  if bt.left.isNil:
    return 1
  else:
    return 1 + bt.left.countNodes() + bt.right.countNodes()

proc createBottomUp(depth: uint): BinaryTree =
  if depth > 0'u:
    return newBinaryTree(createBottomUp(depth - 1), createBottomUp(depth - 1))
  else:
    return newBinaryTree(nil, nil)

when isMainModule:
  if paramCount() != 1:
    stderr.writeLine("Usage: binarytree <depth>")
    quit(1)

  let N =
    try:
      parseUInt(paramStr(1))
    except ValueError:
      stderr.writeLine("Invalid depth argument")
      quit(1)

  var minDepth: uint = 4
  var maxDepth: uint = N
  if (minDepth + 2) > maxDepth:
    maxDepth = minDepth + 2

  block:
    let stretchDepth = maxDepth + 1
    let stretchTree = createBottomUp(stretchDepth)
    echo "stretch tree of depth ", stretchDepth, "\t check: ", stretchTree.countNodes()

  let longLivedTree = createBottomUp(maxDepth)

  var check: uint
  for depth in countup(minDepth, maxDepth, 2):
    let iterations = 1 shl int(maxDepth - depth + minDepth)
    check = 0
    for i in 1 .. iterations:
      let tempTree = createBottomUp(depth)
      check += tempTree.countNodes()
    echo iterations, "\t trees of depth ", depth, "\t check: ", check

  echo "long lived tree of depth ", maxDepth, "\t check: ", longLivedTree.countNodes()

